### PR TITLE
[WIP]netpoll mode

### PR DESCRIPTION
**Pull request by cch123**
https://github.com/mosn/easygo/commit/27b5582505b0975eb95d3299acd3abe5c086f311

1. add field to desc, avoid calling file.Fd which will set the conn to blocking mode

2. remove fd copy in NewDesc, prevent pre close file during GC, before the connection actually close

3. call ReadOnce again until readAgain == false, if there is still data to read when TLS is enabled.

4. clear read timeout timer before doRead, set timer after doRead to enable/disable trigger for read timeout event to send heartbeat request

### CODE REVIEW

File: pkg/network/eventloop.go [hunk 1]
```diff
@@ -32,27 +32,25 @@ var (
...
-	writePool = mosnsync.NewWorkerPool(runtime.NumCPU())
+	readPool  = mosnsync.NewWorkerPool(2 * runtime.NumCPU())
+	writePool = mosnsync.NewWorkerPool(2 * runtime.NumCPU())
 
-	rrCounter                 uint32
-	poolSize                  uint32 = 1 //uint32(runtime.NumCPU())
-	eventLoopPool                    = make([]*eventLoop, poolSize)
-	errEventAlreadyRegistered        = errors.New("event already registered")
+	rrCounter     uint32
+	poolSize      uint32 = 1 //uint32(runtime.NumCPU())
```
Comment (taoyuanyuan) [inline pkg/network/eventloop.go:39]: benchmark to decide poolSize

File: pkg/network/eventloop.go [hunk 2]
```diff
@@ -113,110 +80,40 @@ func (el *eventLoop) registerRead(conn *connection, handler *connEventHandler) e
...
 		}
-
-		el.mu.Lock()
-		delete(el.conn, id)
-		el.mu.Unlock()
 	}
 }
 
-func (el *eventLoop) readWrapper(desc *netpoll.Desc, handler *connEventHandler) func(netpoll.Event) {
+func (el *eventLoop) readWrapper(conn *connection, desc *netpoll.Desc, handler *connEventHandler) func(netpoll.Event) {
```
Comment (taoyuanyuan) [inline pkg/network/eventloop.go:105]: It‘s affect performance that schedule method would block if no worker goroutine is available.

**Comment by cch123** [comment]
GOMAXPROCS=4, use sofaload to benchmark

**Comment by taoyuanyuan** [comment]
add testcase for OnReadTimeout

File: pkg/network/connection.go [hunk 3]
```diff
@@ -222,29 +229,46 @@ func (c *connection) attachEventLoop(lctx context.Context) {
...
 						return true
 					}
 
-					if err == io.EOF {
-						if log.DefaultLogger.GetLogLevel() >= log.DEBUG {
-							log.DefaultLogger.Debugf("[network] [event loop] [onRead] Error on read. Connection = %d, Remote Address = %s, err = %s",
-								c.id, c.RemoteAddr().String(), err)
+					// create a new timer and bind it to connection
+					c.readTimeoutTimer = utils.NewTimer(buffer.ConnReadTimeout, func() {
+						for _, cb := range c.connCallbacks {
```
Comment (taoyuanyuan) [inline pkg/network/connection.go:243]: readbuffer的缩容逻辑

File: pkg/network/connection.go [hunk 4]
```diff
@@ -204,31 +207,68 @@ func (c *connection) attachEventLoop(lctx context.Context) {
 	err := c.eventLoop.registerRead(c, &connEventHandler{
```
Comment (taoyuanyuan) [inline pkg/network/connection.go:207]: 你需要在注册的时候就设置定时器，如果没有请求，你根本没办法设置定时器

File: pkg/network/connection.go [hunk 5]
```diff
@@ -204,31 +207,68 @@ func (c *connection) attachEventLoop(lctx context.Context) {
...
 					}
 
-					if err == io.EOF {
-						if log.DefaultLogger.GetLogLevel() >= log.DEBUG {
-							log.DefaultLogger.Debugf("[network] [event loop] [onRead] Error on read. Connection = %d, Remote Address = %s, err = %s",
-								c.id, c.RemoteAddr().String(), err)
+					// create a new timer and bind it to connection
+					c.readTimeoutTimer = utils.NewTimer(buffer.ConnReadTimeout, func() {
+						for _, cb := range c.connCallbacks {
+							cb.OnEvent(api.OnReadTimeout) // run read timeout callback, for keep alive if configured
```
Comment (taoyuanyuan) [inline pkg/network/connection.go:244]: 第二次怎么设置心跳的定时器

File: pkg/network/connection.go (hunk 4 above)
Comment (cch123) [reply pkg/network/connection.go:207]: 有道理

File: pkg/network/connection.go [hunk 6]
```diff
@@ -204,31 +207,54 @@ func (c *connection) attachEventLoop(lctx context.Context) {
 	err := c.eventLoop.registerRead(c, &connEventHandler{
 		onRead: func() bool {
 			if c.readEnabled {
-				err := c.doRead()
+				c.readTimeoutTimer.Stop()
```
Comment (taoyuanyuan) [inline pkg/network/connection.go:210]: maybe panic.

File: pkg/network/connection.go [hunk 7]
```diff
@@ -239,15 +265,24 @@ func (c *connection) attachEventLoop(lctx context.Context) {
...
 	})
 
 	if err != nil {
 		log.DefaultLogger.Errorf("[network] [event loop] [register] conn %d register read failed:%s", c.id, err.Error())
+		c.Close(api.NoFlush, api.LocalClose)
+		return
 	}
+
+	// create a new timer and bind it to connection
+	c.readTimeoutTimer = utils.NewTimer(buffer.ConnReadTimeout, func() {
```
Comment (taoyuanyuan) [inline pkg/network/connection.go:281]: 应该放到最前面

**Comment by cch123** [comment]
should be merged to develop branch, so deprecate this PR

**Pull request closed without merging**